File: day-04/day-04.py
with open('data.txt', 'r') as r:
    grid = r.read().splitlines()
    p1 = 0
    # horizontal
    for row in grid:
        p1 += row.count("XMAS") + row.count("SAMX")

    # vertical
    for col in range(len(grid[0])):
        col_str = "".join(row[col] for row in grid)
        p1 += col_str.count("XMAS") + col_str.count("SAMX")
    
    # diagonal
    for i in range(len(grid) - 3):
        for j in range(len(grid[0]) - 3):
            main = grid[i][j] + grid[i + 1][j + 1] + grid[i + 2][j + 2] + grid[i + 3][j + 3]
            off = grid[i][j + 3] + grid[i + 1][j + 2] + grid[i + 2][j + 1] + grid[i + 3][j]
            p1 += main.count("XMAS") + off.count("XMAS")
            p1 += main.count("SAMX") + off.count("SAMX")
    
    print("Part 1:", p1) 


    p2 = 0
    for i in range(len(grid) - 2):
        for j in range(len(grid[0]) - 2):
            main = grid[i][j] + grid[i + 1][j + 1] + grid[i + 2][j + 2]
            off = grid[i][j + 2] + grid[i + 1][j + 1] + grid[i + 2][j]
            if main in ["MAS", "SAM"]:
                p2 += off == "MAS" or off == "SAM"
    
    print("Part 2:", p2)

File: day-04/day-04.rs
use::std;

fn main() {
    let input: String = std::fs::read_to_string("day-04/data.txt").unwrap();
    let lines: Vec<&str> = input.lines().collect::<Vec<&str>>();

    let mut part1: usize = 0;

    // horizontal check
    lines.iter().for_each(|line| {
        part1 += line.matches("XMAS").count() + line.matches("SAMX").count();
    });

    // vertical check
    for col in 0..lines[0].len() {
        let mut line: String = String::new();
        for row in &lines {
            line.push(row.chars().nth(col).unwrap());
        }
        part1 += line.matches("XMAS").count() + line.matches("SAMX").count();
    }

    // diagonal check
    for i in 0..(lines.len() - 3) {
        for j in 0..(lines[0].len() - 3) {
            let main: String = (0..4).map(|k| lines[i+k].chars().nth(j+k).unwrap()).collect();
            let off: String = (0..4).map(|k| lines[i+k].chars().nth(j+3-k).unwrap()).collect();
            part1 += main.matches("XMAS").count() + main.matches("SAMX").count();
            part1 += off.matches("XMAS").count() + off.matches("SAMX").count();
        }
    }

    println!("Part 1: {}", part1);

    let mut part2: usize = 0;
    for i in 0..(lines.len() - 2) {
        for j in 0..(lines[0].len() - 2) {
            let main: String = (0..3).map(|k| lines[i+k].chars().nth(j+k).unwrap()).collect();
            let off: String = (0..3).map(|k| lines[i+k].chars().nth(j+2-k).unwrap()).collect();
            if main == "MAS" || main == "SAM" {
                part2 += off.matches("MAS").count() + off.matches("SAM").count();
            }
        }
    }
    
    println!("Part 2: {}", part2);
    
    
}